File: 160_Intersection_of_Two_Linked_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int countA = 0, countB = 0;
        ListNode *a = headA;
        ListNode *b = headB;
        while (a || b) {
            if (a) {
                a = a -> next;
                countA++;
            }
            if (b) {
                b = b -> next;
                countB++;
            }
        }
        int larger = max(countA, countB);
        int smaller = 0;
        ListNode* l;
        ListNode* s;
        if (larger == countA) {
            l = headA;
            s = headB;
            smaller = countB;
        }
        else {
            l = headB;
            s = headA;
            smaller = countA;
        }
        for (int i = larger - smaller; i > 0; i--) l = l -> next;
        while (l && s) {
            if (l == s) return l;
            l = l -> next;
            s = s -> next;
        }
        return NULL;
    }
};

File: 160_Intersection_of_Two_Linked_Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        cntA = cntB = 0
        itrA, itrB = headA, headB
        while itrA or itrB:
            if itrA:
                cntA += 1
                itrA = itrA.next
            if itrB:
                cntB += 1
                itrB = itrB.next
        
        large, small = max(cntA, cntB), 0
        if large == cntA:
            itrL, itrS, small = headA, headB, cntB
        else:
            itrL, itrS, small = headB, headA, cntA
        for i in range(large - small):
            itrL = itrL.next
        while itrL and itrS:
            if itrL == itrS:
                return itrL
            itrL = itrL.next
            itrS = itrS.next
        return None
